src/common/qat.c

```diff
@@ -0,0 +1,537 @@
+    {
+        numInstances = MAX_INSTANCES;
+    }
+    if ((status == CPA_STATUS_SUCCESS) && (numInstances > 0))
+    {
+        status = cpaCyGetInstances(numInstances, cyInstHandles);
+        if (status == CPA_STATUS_SUCCESS)
+            *pCyInstHandle = cyInstHandles[0];
+    }
+
+    if (0 == numInstances)
+    {
+        printf("No QAT instances found.\n");
+    }
+}
+
+static void sal_polling(CpaInstanceHandle cyInstHandle)
+{
+    gPollingCy = 1;
+    while (gPollingCy)
+    {
+        icp_sal_CyPollInstance(cyInstHandle, 0);
+        OS_SLEEP(10);
+    }
+
+    threadExit();
+}
+
+static void startPolling(CpaInstanceHandle cyInstHandle)
+{
+    CpaInstanceInfo2 info2 = {0};
+    CpaStatus status = CPA_STATUS_SUCCESS;
+
+    status = cpaCyInstanceGetInfo2(cyInstHandle, &info2);
+    if ((status == CPA_STATUS_SUCCESS) && (info2.isPolled == CPA_TRUE))
+    {
+        /* Start thread to poll instance */
+        threadCreate(&gPollingThread, sal_polling, cyInstHandle);
+    }
+}
+
+static void stopPolling(void)
+{
+    gPollingCy = 0;
+    OS_SLEEP(10);
+}
+
+/*
+ * Callback function when operation is completed.
+ */
+static void symCallback(void *pCallbackTag,
+                        CpaStatus status,
+                        const CpaCySymOp operationType,
+                        void *pOpData,
+                        CpaBufferList *pDstBuffer,
+                        CpaBoolean verifyResult)
+{
+    if (NULL != pCallbackTag)
+    {
+        /** indicate that the function has been called*/
+        COMPLETE((struct COMPLETION_STRUCT *)pCallbackTag);
+    }
+}
+
+CpaStatus qat_hash_init(CpaInstanceHandle *cyInstHandle, 
+                        CpaCySymSessionCtx *sessionCtx,
+						CpaCySymHashAlgorithm hashAlg, 
+                        Cpa32U digestResultLenInBytes)
+{
+    CpaStatus status = CPA_STATUS_SUCCESS;
+    Cpa32U sessionCtxSize = 0;
+    CpaCySymSessionSetupData sessionSetupData = {0};
+
+    status = qaeMemInit();
+    if (CPA_STATUS_SUCCESS != status)
+    {
+        printf("Failed to initialize memory driver\n");
+        return 0;
+    }
+
+    status = icp_sal_userStartMultiProcess("SSL", CPA_FALSE);
+    if (CPA_STATUS_SUCCESS != status)
+    {
+        printf("Failed to start user process SSL\n");
+        qaeMemDestroy();
+        return 0;
+    }
+    
+    getInstance(cyInstHandle);
+    if (*cyInstHandle == NULL)
+    {
+        return CPA_STATUS_FAIL;
+    }
+
+    status = cpaCyStartInstance(*cyInstHandle);
+
+    if (CPA_STATUS_SUCCESS == status)
+    {
+        /*
+         * Set the address translation function for the instance
+         */
+        status = cpaCySetAddressTranslation(*cyInstHandle, virtToPhys);
+    }
+
+    if (CPA_STATUS_SUCCESS == status)
+    {
+        /*
+         * If the instance is polled start the polling thread. Note that
+         * how the polling is done is implementation-dependent.
+         */
+        startPolling(*cyInstHandle);

```

Comment:
This should really be integrated with Argobots and not start any thread. I would advise to discuss this with @NiuYawei.

Reply:
I'd like to have this initial patch landed as baseline. We can discuss how to optimize and adapt to daos better, but it takes time to evaluate and verify, I think the daos team is in the better position to do it.
What do you think?

Reply:
OK, then let's do a synchronous version for now where the caller is just polling for compression and w/o creating a background thread. Is that doable?

Reply:
Good point, we can do it.